Refactor simulation initialisation

AstroKriel opened the pull request:
Boilerplate initialization code has been moved to the AMRSimulation class. Two functions are exposed to the user, first, which allows the user to perform a computation before initialization (for example to read in a table), and a second is called while the initialization code iterates over the simulation domain. These functions have empty implementations by default, so the user will need to implement them when relevant.

AstroKriel committed:
fix init-ing bug: now init-ing all state varaibles (for running in debug mode)

AstroKriel committed:
Merge branch 'BenWibking:development' into development

AstroKriel committed:
Merge branch 'development' of github.com:AstroKriel/Quokka into development

AstroKriel committed:
Merge branch 'BenWibking:development' into development

AstroKriel committed:
Delete highmach_output.csv

AstroKriel committed:
Merge branch 'BenWibking:development' into development

AstroKriel committed:
removed extraneous files

AstroKriel committed:
created a container for storing information relevant when initialising different grids (i.e., cell- or face-centred)

AstroKriel committed:
provide empty implementations for funcs evaluated in the init step. the user should provide an implementation for each in the problem file if required.

AstroKriel committed:
relocate boiler-plate code for simulation initialisation to the parent simulation class, with calls to user-defined funcs.

AstroKriel committed:
implement ini funcs for each problem file (in progress)

AstroKriel committed:
ignore folders and subcontents for testing

AstroKriel committed:
Fix P -> Pgas in Hydro_System assert

BenWibking committed:
Merge branch 'development' into refactor-sim-init

BenWibking commented on src/grid.hpp:
```diff
@@ -0,0 +1,25 @@
+#include <AMReX.H>
```
This needs an include guard.

AstroKriel replied:
Done.

BenWibking commented on src/grid.hpp:
```diff
@@ -0,0 +1,25 @@
+#include <AMReX_Box.H>
+#include <AMReX_REAL.H>
+
+enum class centering { cc=0, fc, ec };
```
put the enums and the struct inside a `namespace quokka { .. }` block

AstroKriel replied:
Done.

BenWibking commented on src/simulation.hpp:
```diff
@@ -942,6 +969,9 @@ void AMRSimulation<problem_t>::MakeNewLevelFromScratch(
 
+  amrex::IntVect iv(AMREX_D_DECL(19, 0, 5));
+  print_state(state_new_[level], iv);
+
```
Delete this debugging code.

AstroKriel replied:
Ugh, I thought I had already deleted this. Done now (again).

BenWibking reviewed (changes requested):
Looks good overall. The main thing is I think we want to put the `grid` struct inside the `quokka` namespace. This means it will be referenced with `quokka::grid` instead of just `grid`.

AstroKriel committed:
added include gaurds for the header file

AstroKriel committed:
put grid struct and grid enums in a quokka namespace

AstroKriel committed:
remove debugging code

AstroKriel committed:
Merge branch 'refactor-sim-init' of github.com:AstroKriel/Quokka into refactor-sim-init

BenWibking commented:
/azp run

BenWibking commented:
@AstroKriel can you take a look at the build errors?

AstroKriel committed:
add quokka namespace call for grid

AstroKriel committed:
add quokka namespace call for grid in problem files

AstroKriel commented:
> @AstroKriel can you take a look at the build errors?

Yep, fixed the errors. The errors were because I forgot to add the quokka namespace call to all references of the grid-struct. I have checked everything compiles fine now.

BenWibking commented:
/azp run

BenWibking commented:
> > @AstroKriel can you take a look at the build errors?
> 
> Yep, fixed the errors. The errors were because I forgot to add the quokka namespace call to all references of the grid-struct. I have checked everything compiles fine now.

The GPU build is still failing. Can you take a look at that?

AstroKriel committed:
move access of state variables outside of func (i.e., computeWaveSolution) which is potentially evaluated on GPU, but grid_vec[0].array lives on CPU.

AstroKriel commented:
> > > @AstroKriel can you take a look at the build errors?
> > 
> > 
> > Yep, fixed the errors. The errors were because I forgot to add the quokka namespace call to all references of the grid-struct. I have checked everything compiles fine now.
> 
> The GPU build is still failing. Can you take a look at that?

It looks like the build error occurred because the reference to the state array was stored in grid_vec[...].array, and that vector lives on the CPU. There isn't an operator implemented to access vector elements stored on the CPU from the GPU, and that is why the error was being thrown. I fixed this by extracting the reference to the state variable and passing that reference directly to computeWaveSolution(...) instead of accessing the reference via the grid_vec[...] vector in the function call.

BenWibking commented:
> It looks like the build error occurred because the reference to the state array was stored in grid_vec[...].array, and that vector lives on the CPU. There isn't an operator implemented to access vector elements stored on the CPU from the GPU, and that is why the error was being thrown. I fixed this by extracting the reference to the state variable and passing that reference directly to computeWaveSolution(...) instead of accessing the reference via the grid_vec[...] vector in the function call.

Yeah, the fact that `std::vector` is not usable from the GPU requires a lot of workarounds in general. This looks like a reasonable way to do it. I'll re-run the GPU tests and merge if it passes.

BenWibking committed:
Merge branch 'development' into refactor-sim-init

BenWibking commented:
/azp run

BenWibking commented:
@AstroKriel something is still not right. GPU build still fails.

AstroKriel commented:
Ah yep, it looks like the state reference (ie grid_vec[...].array) needs to be extracted outside all the ParFor loops, because when compiled on GPU, the same problem as before occurs. I'll fix this and see if the problem persists / if there are other problems.

AstroKriel committed:
move access of state variables reference outside of ParFor.

AstroKriel committed:
Merge branch 'refactor-sim-init' of github.com:AstroKriel/Quokka into refactor-sim-init

BenWibking commented:
/azp run

BenWibking commented:
@AstroKriel there are GPU build failures in the HydroBlast3D problem.

Commit:
create a copy of global variables so  if the program is compiled to run on the GPU, then the GPU  will be able to access variables.

BenWibking commented:
/azp run
